src/server/builder.ts

```diff
@@ -0,0 +1,405 @@
+         */
+        public updateLastEmittedContentSignature(builder: Builder) {
+            this.lastEmittedContentSignature = this.getScriptInfo(builder).getLatestVersion();
+        }
+
+        public checkIfShapeSignatureChanged(builder: Builder) {
+            const sourceFile = builder.project.getSourceFile(this.fileName);
+            if (!sourceFile) {
+                return true;
+            }
+
+            const lastSignature = this.lastCheckedShapeSignature;
+            if (sourceFile.isDeclarationFile) {
+                this.lastCheckedShapeSignature = this.computeHash(sourceFile.text);
+            }
+            const emitOutput = builder.project.getLanguageService(/*ensureSynchronized*/true).getEmitOutput(this.fileName, /*emitDeclarationsOnly*/ true);
+            for (const file of emitOutput.outputFiles) {
+                if (/\.d\.ts$/.test(file.name)) {
+                    this.lastCheckedShapeSignature = this.computeHash(file.text);
+                }
+            }
+            return !lastSignature || this.lastCheckedShapeSignature !== lastSignature;
+        }
+
+        public isContentSignatureChangedSinceLastEmit(builder: Builder): boolean {
+            const info = this.getScriptInfo(builder);
+            return !this.lastEmittedContentSignature || info.getLatestVersion() !== this.lastEmittedContentSignature;
+        }
+    }
+
+
+    export interface Builder {
+        readonly project: Project;
+        getFilesAffectedBy(fileName: string): string[];
+        onProjectUpdateGraph(): void;
+        emitFile(fileName: string, writeFile: (path: string, data: string, writeByteOrderMark?: boolean) => void, forced?: boolean): boolean;
+    }
+
+    class NullBuilder implements Builder {
+        constructor(readonly project: Project) {
+        }
+
+        getFilesAffectedBy(fileName: string): string[] {
+            return undefined;
+        }
+
+        onProjectUpdateGraph() {
+        }
+
+        updateEmittedFileContentSignature(fileName: string): void {
+        }
+
+        emitFile(fileName: string, writeFile: (path: string, data: string, writeByteOrderMark?: boolean) => void, forced?: boolean): boolean {
+            return true;
+        }
+    }
+
+    abstract class AbstractBuilder<T extends BuilderFileInfo> implements Builder {
+
+        protected fileInfos: Map<T> = {};
+
+        constructor(public readonly project: Project) {
+        }
+
+        protected getFileInfo(fileName: string): T {
+            return this.fileInfos[fileName];
+        }
+        abstract getFilesAffectedBy(fileName: string): string[];
+        abstract onProjectUpdateGraph(): void;
+
+        emitFile(fileName: string, writeFile: (path: string, data: string, writeByteOrderMark?: boolean) => void, forced?: boolean): boolean {
+            const fileInfo = this.getFileInfo(fileName);
+            if (!fileInfo || (!forced && !fileInfo.isContentSignatureChangedSinceLastEmit(this))) {

```

Comment:
nit: IMO it is a bit hard to read since there are too many negations. I would rewrite it as:
```ts
const shouldEmit = fileInfo && (forced || fileInfo.isContentSignatureChangedSinceLastEmit(this));
if (!shouldEmit) {
    return false;
}
```